core/trino-main/src/main/java/io/trino/cost/CachingTableStatsProvider.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.cost;
+
+import io.trino.Session;
+import io.trino.metadata.Metadata;
+import io.trino.metadata.TableHandle;
+import io.trino.spi.statistics.TableStatistics;
+
+import java.util.Map;
+import java.util.WeakHashMap;
+
+import static java.util.Objects.requireNonNull;
+
+public class CachingTableStatsProvider
+        implements TableStatsProvider
+{
+    private final Metadata metadata;
+    private final Session session;
+
+    private final Map<TableHandle, TableStatistics> cache = new WeakHashMap<>();
```

Comment:
nit: why do you need WeakHashMap if CTSP is created per query. Do you expect that `TableHandle` objects will be GCed often while query is running. It is possible for sure but I am not sure if often in practice.

Reply:
During optimizations we can create large number of TableHandle objects (eg subsequent pushdowns) and old TableHandle objects may not be reachable in the plan. I don't think "large number" would be a common situation (there isn't an unlimited number of pushdown opportunities for a given query), but i still don't think we should be using strong references here.